File: rpi_lepton_video/16to8/src/16to8.cpp
#include "16to8.h"
#include <string>

#include<stdio.h>
#include<cstdlib>
#include<iostream>
#include<string.h>
#include<fstream>
#include<dirent.h>

using namespace std;

void processDir(string in, string out, int num_files) {
    
//    cv::VideoWriter vid(out.c_str(), CV_FOURCC('X','V','I','D'), 25, cv::Size(80,60)); 

    for (int i=1; i <= num_files; i++) {
        char fname[50];
        sprintf(fname, "/img_%06d.png", i);
        
        cv::Mat image = cv::imread(in + "/" + fname, -1);
//        cout << fname << "\n";
        
        cv::Mat output;
        double minVal, maxVal;
        minMaxLoc(image, &minVal, &maxVal);
        image.convertTo(output, 8, 255.0/(maxVal - minVal), -minVal * 255.0/(maxVal - minVal));

//        vid.write(output);

        imwrite(out + "/" + fname, output);
    }
}

int main(int argc, char** argv )
{
    if ( argc != 4 )
    {
        printf("usage: 16to8 <input_dir> <output_dir> <num_files>\n");
        return -1;
    }

    string input_dir = argv[1];
    string output_dir = argv[2];
    int num_files = atoi(argv[3]);

    processDir(input_dir, output_dir, num_files);

    return 0;
}

File: rpi_lepton_video/16to8/src/16to8.py
import os
import sys

import cv2
import numpy as np
import png

def find_max_min(img):
  rows, cols = img.shape
  max_pxl = 0
  min_pxl = pow(2,16)-1

  for i in range(rows):
    for j in range(cols):
      if img[i,j] > max_pxl:
        max_pxl = img[i,j]
      if img[i,j] < min_pxl:
        min_pxl = img[i,j]

  return (max_pxl, min_pxl)

def scale_img(img, max_pxl, min_pxl):
  rows, cols = img.shape
  downscaled = np.zeros((rows, cols), dtype=np.uint8)
  diff = max_pxl - min_pxl
  scale = diff / 255
  for i in range(rows):
    for j in range(cols):
      downscaled[i,j] = (img[i,j] - min_pxl) / scale
      if downscaled[i,j] > 255:
        print downscaled[i,j]
  return downscaled

if len(sys.argv) < 3:
  print "Usage: 16to8.py <input_dir> <output_dir>"
  sys.exit()
  
def copy_img(in_img, out_img):
  rows, cols = in_img.shape

  for i in range(rows):
    for j in range(cols):
        out_img[i,j] = img[i,j]

  return out_img

input_dir = sys.argv[1]
output_avi = sys.argv[2]

fourcc = cv2.cv.CV_FOURCC('X','V','I','D')
vid = cv2.VideoWriter(output_avi, fourcc, 27, (80,60))

for in_file in os.listdir(input_dir):
  img = cv2.imread(os.path.join(input_dir, in_file), -1)

#  max_pxl, min_pxl = find_max_min(img)
#  print max_pxl, min_pxl
  img *= 255.0/img.max()
  max_pxl, min_pxl = find_max_min(img)
  print max_pxl, min_pxl
  
  downscaled = np.array((60, 80, 1), dtype=np.uint8)
  downscaled = copy_img(img, downscaled)

  cv2.imshow('window', img)
  cv2.imwrite(in_file + "_n")
  cv2.waitKey(0)
  vid.write(img)

vid.release()
